File: tools/tal_support/as274_fc/C/t8.c
#include <math.h>
#include <stdio.h>
#include "as274c.h"
#include "misc.h"

int main()
{
    const int np=8, nrbar=28;
    int error, icase, i, *vorder, *list, *lindep, new;
    double *d, *rbar, *thetab, *rss, *tol, sserr, *x, y, *beta;

    vorder = ivector(0, np-1);
    list = ivector(0, np-1);
    d = dvector(0, np-1);
    rbar = dvector(0, nrbar-1);
    thetab = dvector(0, np-1);
    rss = dvector(0, np-1);
    tol = dvector(0, np-1);
    x = dvector(0, np-1);
    beta = dvector(0, np-1);
    lindep = ivector(0, np-1);
    
    randc.ix = randc.iy = randc.iz = 777;
    error = clear(np, nrbar, d, rbar, thetab, &sserr);
    if (error) printf("clear() returned %d\n", error);

    for (icase = 0; icase<5; icase++) {
        for (i=0; i<4; i++) x[i] = ranwm();
        x[4] = x[0] + x[1];
        x[5] = x[1] + x[2];
        x[6] = x[2] + x[3];
        x[7] = x[3] + x[0];
        y = x[4] + x[6];
        error = includ(np, nrbar, 1, x, y, d, rbar, thetab, &sserr);
        if (error) printf("At row %d, includ() returned %d\n", icase, error);
    }

    /* Now look at factorisation */
    putdvec("D", d, 0, 7);
    pr_utdm_v(rbar, 7, 11, 3);

    error = tolset(np, nrbar, d, rbar, tol);
    if (error) printf("tolset() returned %d\n", error);
    error = sing(np, nrbar, d, rbar, thetab, &sserr, tol, lindep);
    if (error) printf("sing() returned %d\n", error);
    printf("lindep vector is: ");
    for (i=0; i<np; i++) printf("%d ", lindep[i]); printf("\n");

    error = ss(np, d, thetab, &sserr, rss);
    if (error) printf("ss() returned %d\n", error);
    error = regcf(np, nrbar, d, rbar, thetab, tol, beta, 4);
    printf("betas: %f %f %f %f\n", beta[0], beta[1], beta[2], beta[3]);

    /* reorder vars as x5..x8, x1..x4.  Note xt+x7 == x6+x8 */
    new = 4;
    for (i=0; i<8; i++) {
        vorder[i] = i;
        list[i] = new;
        new = i == 3 ? 0 : new+1;
    }
    error = reordr(np, nrbar, vorder, d, rbar, thetab, rss, tol, list, 4, 1);
    if (error) printf("reordr() returned %d\n", error);

    /* Look at factorisation again: */
    putdvec("D", d, 0, 7);
    pr_utdm_v(rbar, 7, 11, 3);
    error = sing(np, nrbar, d, rbar, thetab, &sserr, tol, lindep);
    if (error) printf("sing() returned %d\n", error);
    printf("lindep vector is: ");
    for (i=0; i<np; i++) printf("%d ", lindep[i]); printf("\n");

    error = regcf(np, nrbar, d, rbar, thetab, tol, beta, 4);
    printf("betas: %f %f %f %f\n", beta[0], beta[1], beta[2], beta[3]);

    return 0;
}

File: tools/tal_support/as274_fc/F/t8.f

      program test8
c
c     Test handling of a case with more variables than cases.
c     X1, .., X4 from random number generator; Y = X1 + X2 + X3 + X4
c     X5 = X1 + X2
c     X6 =      X2 + X3
c     X7 =           X3 + X4
c     X8 = X1           + X4   Hence Y = (X5 + X6 + X7 + X8)/2
c
c     ncases = 5
c
c     The matrix RBAR should look like:
c		(1)   1   2   3   4   5   6   7
c		     (1)  8   9  10  11  12  13
c			 (1) 14  15  16  17  18
c			     (1) 19  20  21  22
c				 (1) 23  24  25
c                                    (1) 26  27
c                                        (1) 28
c     where (1) indicates an implicit 1.0 and the other numbers are the
c     locations within array RBAR.   Elements 26-28 should be zero.
c
      integer np, nrbar
      parameter (np=8, nrbar=28)
      integer ix, iy, iz, ier, case, i, vorder(np), list(np)
      real rand
      double precision d(np), rbar(28), thetab(np), rss(np),
     +    wt, tol(np), sserr, x(8), y, work(np), beta(np)
      logical lindep(np)
      common /randc/ ix, iy, iz
      data wt/1.0d0/

      ix = 777
      iy = 777
      iz = 777
      call clear(np, nrbar, d, rbar, thetab, sserr, ier)
c
c     Generate 5 lines of data
c
      do 20 case = 1, 5
	do 10 i = 1, 4
   10   x(i) = rand()
	x(5) = x(1) + x(2)
	x(6) = x(2) + x(3)
	x(7) = x(3) + x(4)
	x(8) = x(4) + x(1)
	y = x(5) + x(7)
	call includ(np, nrbar, wt, x, y, d, rbar, thetab,
     +		sserr, ier)
   20 continue
c
c     Now look at the factorization
c
      write(*, 900) d
  900 format(' D = ', 8f9.5)
      write(*, 910)(rbar(i),i=1,7)
  910 format(' RBAR = '/ 1x, 'Row 1: (1) ', 7f9.5)
      write(*, 920)(rbar(i),i=8,13)
  920 format(1x, 'Row 2:', 9x, ' (1) ', 6f9.5)
      write(*, 930)(rbar(i),i=14,18)
  930 format(1x, 'Row 3:', 18x, ' (1) ', 5f9.5)
      write(*, 940)(rbar(i),i=19,22)
  940 format(1x, 'Row 4:', 27x, ' (1) ', 4f9.5)
      write(*, 950)(rbar(i),i=23,25)
  950 format(1x, 'Row 5:', 36x, ' (1) ', 3f9.5)
      write(*, 960)(rbar(i),i=26,28)
  960 format(1x, 'Row 6:', 45x, ' (1) ', 2f9.5/
     +       1x, 'Row 7:', 54x, ' (1) ', f9.5)

      call tolset(np, nrbar, d, rbar, tol, work, ier)
      call sing(np, nrbar, d, rbar, thetab, sserr, tol,
     +		lindep, work, ier)
      write(*, *)' IER = ', ier
      write(*, *)' LINDEP = ', lindep
      call ss(np, d, thetab, sserr, rss, ier)
c
c     Calculate regression coefficients for first 4 variables.
c
      call regcf(np, nrbar, d, rbar, thetab, tol, beta,
     +		4, ier)
      write(*, 970) (beta(i),i=1,4)
  970 format(' Regression coefficients:'/ 1x, 4f10.5/)
c
c     Re-order variables as X5 .. X8, X1 .. X4
c     N.B.  X5 + X7 = X6 + X8
c
      new = 5
      do 30 i = 1, 8
	vorder(i) = i
	list(i) = new
	if (i .eq. 4) then
	  new = 1
	else
	  new = new + 1
	end if
   30 continue
      call reordr(np, nrbar, vorder, d, rbar, thetab, rss,
     +		tol, list, 4, 1, ier)
c
c     Now look at the factorization again
c
      write(*, 900) d
      write(*, 910)(rbar(i),i=1,7)
      write(*, 920)(rbar(i),i=8,13)
      write(*, 930)(rbar(i),i=14,18)
      write(*, 940)(rbar(i),i=19,22)
      write(*, 950)(rbar(i),i=23,25)
      write(*, 960)(rbar(i),i=26,28)
      call sing(np, nrbar, d, rbar, thetab, sserr, tol,
     +		lindep, work, ier)
      write(*, *)' IER = ', ier
      write(*, *)' LINDEP = ', lindep
c
c     Calculate regression coefficients for first 4 variables.
c
      call regcf(np, nrbar, d, rbar, thetab, tol, beta, 4, ier)
      write(*, 970) (beta(i),i=1,4)

      end
